// typesafe-utils/lib/esm/sorting/sorting.js
const sortASC = (a, b) => {
    if (a === b)
        return 0;
    if (a == null)
        return -1;
    if (b == null)
        return 1;
    return a < b ? -1 : 1;
};
export const sortNumberASC = (a, b) => sortASC(a, b);
export const sortNumberPropertyASC = (property) => ({ [property]: a }, { [property]: b }) => sortNumberASC(a, b);
export const sortNumberDESC = (a, b) => sortNumberASC(b, a);
export const sortNumberPropertyDESC = (property) => ({ [property]: a }, { [property]: b }) => sortNumberDESC(a, b);
export const sortStringASC = (a, b) => sortASC(a === null || a === void 0 ? void 0 : a.toLowerCase(), b === null || b === void 0 ? void 0 : b.toLowerCase());
export const sortStringPropertyASC = (property) => ({ [property]: a }, { [property]: b }) => sortStringASC(a, b);
export const sortStringDESC = (a, b) => sortStringASC(b, a);
export const sortStringPropertyDESC = (property) => ({ [property]: a }, { [property]: b }) => sortStringDESC(a, b);
export const sortDateASC = (a, b) => sortASC(a === null || a === void 0 ? void 0 : a.getTime(), b === null || b === void 0 ? void 0 : b.getTime());
export const sortDatePropertyASC = (property) => ({ [property]: a }, { [property]: b }) => sortASC(a === null || a === void 0 ? void 0 : a.getTime(), b === null || b === void 0 ? void 0 : b.getTime());
export const sortDateDESC = (a, b) => sortDateASC(b, a);
export const sortDatePropertyDESC = (property) => ({ [property]: a }, { [property]: b }) => sortDateDESC(a, b);
