// typesafe-utils/lib/cjs/sorting/sorting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sortDatePropertyDESC = exports.sortDateDESC = exports.sortDatePropertyASC = exports.sortDateASC = exports.sortStringPropertyDESC = exports.sortStringDESC = exports.sortStringPropertyASC = exports.sortStringASC = exports.sortNumberPropertyDESC = exports.sortNumberDESC = exports.sortNumberPropertyASC = exports.sortNumberASC = void 0;
const sortASC = (a, b) => {
    if (a === b)
        return 0;
    if (a == null)
        return -1;
    if (b == null)
        return 1;
    return a < b ? -1 : 1;
};
const sortNumberASC = (a, b) => sortASC(a, b);
exports.sortNumberASC = sortNumberASC;
const sortNumberPropertyASC = (property) => ({ [property]: a }, { [property]: b }) => exports.sortNumberASC(a, b);
exports.sortNumberPropertyASC = sortNumberPropertyASC;
const sortNumberDESC = (a, b) => exports.sortNumberASC(b, a);
exports.sortNumberDESC = sortNumberDESC;
const sortNumberPropertyDESC = (property) => ({ [property]: a }, { [property]: b }) => exports.sortNumberDESC(a, b);
exports.sortNumberPropertyDESC = sortNumberPropertyDESC;
const sortStringASC = (a, b) => sortASC(a === null || a === void 0 ? void 0 : a.toLowerCase(), b === null || b === void 0 ? void 0 : b.toLowerCase());
exports.sortStringASC = sortStringASC;
const sortStringPropertyASC = (property) => ({ [property]: a }, { [property]: b }) => exports.sortStringASC(a, b);
exports.sortStringPropertyASC = sortStringPropertyASC;
const sortStringDESC = (a, b) => exports.sortStringASC(b, a);
exports.sortStringDESC = sortStringDESC;
const sortStringPropertyDESC = (property) => ({ [property]: a }, { [property]: b }) => exports.sortStringDESC(a, b);
exports.sortStringPropertyDESC = sortStringPropertyDESC;
const sortDateASC = (a, b) => sortASC(a === null || a === void 0 ? void 0 : a.getTime(), b === null || b === void 0 ? void 0 : b.getTime());
exports.sortDateASC = sortDateASC;
const sortDatePropertyASC = (property) => ({ [property]: a }, { [property]: b }) => sortASC(a === null || a === void 0 ? void 0 : a.getTime(), b === null || b === void 0 ? void 0 : b.getTime());
exports.sortDatePropertyASC = sortDatePropertyASC;
const sortDateDESC = (a, b) => exports.sortDateASC(b, a);
exports.sortDateDESC = sortDateDESC;
const sortDatePropertyDESC = (property) => ({ [property]: a }, { [property]: b }) => exports.sortDateDESC(a, b);
exports.sortDatePropertyDESC = sortDatePropertyDESC;
