// typesafe-utils/lib/cjs/isUndefined/isUndefined.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.arePropertiesNotUndefined = exports.arePropertiesUndefined = exports.isPropertyNotUndefined = exports.isPropertyUndefined = exports.isNotUndefined = exports.isUndefined = void 0;
const isUndefined = (value) => value === undefined;
exports.isUndefined = isUndefined;
const isNotUndefined = (value) => value !== undefined;
exports.isNotUndefined = isNotUndefined;
const isPropertyUndefined = (property) => (value) => value[property] === undefined;
exports.isPropertyUndefined = isPropertyUndefined;
const isPropertyNotUndefined = (property) => (value) => value[property] !== undefined;
exports.isPropertyNotUndefined = isPropertyNotUndefined;
const arePropertiesUndefined = (...properties) => (value) => properties.every((property) => value[property] === undefined);
exports.arePropertiesUndefined = arePropertiesUndefined;
const arePropertiesNotUndefined = (...properties) => (value) => properties.every((property) => value[property] !== undefined);
exports.arePropertiesNotUndefined = arePropertiesNotUndefined;
